### PR TITLE
HDDS-1829 On OM reload/restart OmMetrics#numKeys should be updated

### PR DESCRIPTION
**Pull request by smengcl**

### CODE REVIEW

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java [hunk 1]
```diff
@@ -183,4 +183,14 @@ public String getName() throws IOException {
...
     // Nothing do for a Column Family.
   }
+
+  @Override
+  public long getEstimatedKeyCount() throws IOException {
+    try {
+      return db.getLongProperty(handle, "rocksdb.estimate-num-keys");
+    } catch (RocksDBException e) {
+      throw new IOException(
+          "Failed to get estimated key count of table.");
```
Comment (smengcl) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java]: Does this exception handling look good to you?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java [hunk 2]
```diff
@@ -205,6 +205,16 @@ public String getName() throws IOException {
...
   }
 
+  @Override
+  public long getEstimatedKeyCount() throws IOException {
+    if (rawTable instanceof RDBTable) {
+      return rawTable.getEstimatedKeyCount();
+    }
+    throw new IllegalArgumentException(
+        "Unsupported operation getEstimatedKeyCount() on table type " +
+            rawTable.getClass().getCanonicalName());
```
Comment (smengcl) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java]: Does IllegalArgumentException look good here?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java (hunk 1 above)
Comment (arp7) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java]: If we throw an exception here, Ratis will get a failure via the future returned from `notifyInstallSnapshotFromLeader`. I am not sure what the impact of that is. Will Ratis terminate the process? cc @hanishakoneru 

I'd say for now this is okay. I filed https://issues.apache.org/jira/browse/HDDS-1862 to investigate further.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java [hunk 3]
```diff
@@ -183,4 +183,14 @@ public String getName() throws IOException {
...
     // Nothing do for a Column Family.
   }
+
+  @Override
+  public long getEstimatedKeyCount() throws IOException {
+    try {
+      return db.getLongProperty(handle, "rocksdb.estimate-num-keys");
+    } catch (RocksDBException e) {
+      throw new IOException(
+          "Failed to get estimated key count of the table.");
```
Comment (bharatviswa504) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java]: Minor Nit: Can we add the table name here.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java (hunk 2 above)
Comment (bharatviswa504) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java]: I don't think we need a check here, as the underlying table for TypedTable is rocksdbTable.

**Comment by hadoop-yetus** [comment]
:broken_heart: **-1 overall**






| Vote | Subsystem | Runtime | Comment |
|:----:|----------:|--------:|:--------|
| 0 | reexec | 99 | Docker mode activated. |
||| _ Prechecks _ |
| +1 | dupname | 0 | No case conflicting files found. |
| +1 | @author | 0 | The patch does not contain any @author tags. |
| +1 | test4tests | 0 | The patch appears to include 2 new or modified test files. |
||| _ trunk Compile Tests _ |
| 0 | mvndep | 75 | Maven dependency ordering for branch |
| +1 | mvninstall | 658 | trunk passed |
| +1 | compile | 396 | trunk passed |
| +1 | checkstyle | 77 | trunk passed |
| +1 | mvnsite | 0 | trunk passed |
| +1 | shadedclient | 905 | branch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 176 | trunk passed |
| 0 | spotbugs | 483 | Used deprecated FindBugs config; considering switching to SpotBugs. |
| +1 | findbugs | 723 | trunk passed |
||| _ Patch Compile Tests _ |
| 0 | mvndep | 33 | Maven dependency ordering for patch |
| -1 | mvninstall | 144 | hadoop-ozone in the patch failed. |
| -1 | compile | 51 | hadoop-ozone in the patch failed. |
| -1 | javac | 51 | hadoop-ozone in the patch failed. |
| -0 | checkstyle | 39 | hadoop-hdds: The patch generated 2 new + 0 unchanged - 0 fixed = 2 total (was 0) |
| +1 | mvnsite | 0 | the patch passed |
| +1 | whitespace | 0 | The patch has no whitespace issues. |
| +1 | shadedclient | 732 | patch has no errors when building and testing our client artifacts. |
| +1 | javadoc | 155 | the patch passed |
| -1 | findbugs | 102 | hadoop-ozone in the patch failed. |
||| _ Other Tests _ |
| -1 | unit | 357 | hadoop-hdds in the patch failed. |
| -1 | unit | 121 | hadoop-ozone in the patch failed. |
| +1 | asflicense | 38 | The patch does not generate ASF License warnings. |
| | | 5489 | |


| Subsystem | Report/Notes |
|----------:|:-------------|
| Docker | Client=19.03.1 Server=19.03.1 base: https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/Dockerfile |
| GITHUB PR | https://github.com/apache/hadoop/pull/1164 |
| Optional Tests | dupname asflicense compile javac javadoc mvninstall mvnsite unit shadedclient findbugs checkstyle |
| uname | Linux c3987cf80d2b 4.15.0-48-generic #51-Ubuntu SMP Wed Apr 3 08:28:49 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux |
| Build tool | maven |
| Personality | personality/hadoop.sh |
| git revision | trunk / aebac6d |
| Default Java | 1.8.0_212 |
| mvninstall | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-mvninstall-hadoop-ozone.txt |
| compile | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-compile-hadoop-ozone.txt |
| javac | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-compile-hadoop-ozone.txt |
| checkstyle | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/diff-checkstyle-hadoop-hdds.txt |
| findbugs | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-findbugs-hadoop-ozone.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-unit-hadoop-hdds.txt |
| unit | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/artifact/out/patch-unit-hadoop-ozone.txt |
|  Test Results | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/testReport/ |
| Max. process+thread count | 377 (vs. ulimit of 5500) |
| modules | C: hadoop-hdds/common hadoop-ozone/common hadoop-ozone/ozone-manager U: . |
| Console output | https://builds.apache.org/job/hadoop-multibranch/job/PR-1164/1/console |
| versions | git=2.7.4 maven=3.3.9 findbugs=3.1.0-RC1 |
| Powered by | Apache Yetus 0.10.0 http://yetus.apache.org |


This message was automatically generated.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java (hunk 2 above)
Comment (smengcl) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/TypedTable.java]: Thanks for that. I've pushed a commit for this.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java (hunk 3 above)
Comment (smengcl) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java]: Sure! Done.

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java (hunk 1 above)
Comment (smengcl) [reply hadoop-hdds/common/src/main/java/org/apache/hadoop/utils/db/RDBTable.java]: Though I'm not sure how the exception would affect Ratis, I found there are a bunch of methods in this class that are converting RocksDBException into IOException as well, like `RDBTable#get`, `RDBTable#delete`, `RDBTable#getName`.

**Comment by bharatviswa504** [comment]
Thank You @arp7 for the review and @smengcl for the contribution.
I will commit this to the trunk.

**Comment by bharatviswa504** [comment]
Reverted this change from the trunk. I see there are some check style issues.

**Comment by bharatviswa504** [comment]
@smengcl can you open a new PR to get CI run.

**Comment by smengcl** [comment]
@bharatviswa504 Sure.

**Comment by smengcl** [comment]
New PR with checkstyle fix: https://github.com/apache/hadoop/pull/1187

Pending CI.

**Pull request merged**